### PR TITLE
Controller

### PR DESCRIPTION
**Pull request by davepl**
Adds an outer main class called Controller that holds the canvas array.  Make the 7777 port an option on the command line.  Various changes to the interfaces and serialization to suport the CRUD api.

### CODE REVIEW

File: canvas.h [hunk 1]
```diff
@@ -59,31 +65,42 @@ class Canvas : public ICanvas
         return _effects;
     }
 
-    vector<unique_ptr<ILEDFeature>>& Features() override
+    vector<ILEDFeature *> Features() override
```
Comment (rbergen) [inline canvas.h:68]: If `unique_ptr` is no longer usable for this scenario, wouldn't it be better to use a `shared_ptr` instead?

File: controller.h [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+        }
+        return canvas->RemoveFeatureById(featureId);
+    }
+
+
+    // LoadSampleCanvases
+    //
+    // Until we have full load and save ability, this function will be used to load sample canvases.
+
+    void LoadSampleCanvases() override
```
Comment (rbergen) [inline controller.h:85]: I think you've made it a lot easier for me to override your Canvas setup with mine. :)

**Review by rbergen — COMMENTED** [review]
I have one question about the use of raw pointers in canvas.h, that I would like to understand before we merge.

Other than that it looks good to me.

Note that I haven't tested this branch myself; I fully trust you did.

**Comment by davepl** [comment]
I want ownership to be retained via unique_ptr in the controller.  I asked ChatGPT and Claude about it, and both argued for raw pointers without any shared allocation, since the lifetime is guaranteed for the call.

At first it struct me as “old” but I think its the right solution for ownership and lifetime.  If you have a counter, I’m all ears!

- Dave

**Comment by rbergen** [comment]
Nope, now I see I misread. I somehow missed that this concerns return
values only; I thought I was looking at parameters. I guess that's the risk
of lunchtime reviews...

I was worrying about the lifecycle management (read: cleanup) of the
features, but that is still managed by the unique_ptr encapsulating the raw
one. As you say.

My apologies for starting a point(er)less (ha ha) conversation on you. :)

Cheers,
Rutger

**Comment by davepl** [comment]
BTW, you can just move on to the PR for the Thursday branch, which should include these changes as well.

**Pull request merged**